// @sudoku/stores/candidates.js
import { writable } from 'svelte/store';

function createCandidates() {
	const candidates = writable({});

	return {
		subscribe: candidates.subscribe,

		add(pos, candidate) {
			candidates.update($candidates => {
				if (!$candidates.hasOwnProperty(pos.x + ',' + pos.y)) {
					$candidates[pos.x + ',' + pos.y] = [candidate];
				} else if ($candidates[pos.x + ',' + pos.y].includes(candidate)) {
					delete $candidates[pos.x + ',' + pos.y][$candidates[pos.x + ',' + pos.y].indexOf(candidate)];
				} else {
					$candidates[pos.x + ',' + pos.y].push(candidate);
				}

				return $candidates;
			});
		},

		setCandidates(pos, list) {
			candidates.update($candidates => {
				const key = pos.x + ',' + pos.y;
				if (!Array.isArray(list) || list.length === 0) {
					delete $candidates[key];
					return $candidates;
				}

				const unique = [...new Set(list)]
					.filter(val => Number.isInteger(val) && val >= 1 && val <= 9)
					.sort((a, b) => a - b);

				$candidates[key] = unique;
				return $candidates;
			});
		},

		clear(pos) {
			candidates.update($candidates => {
				delete $candidates[pos.x + ',' + pos.y];
				return $candidates;
			});
		}
	}
}

export const candidates = createCandidates();
