// @sudoku/game.js
import { cursor } from './stores/cursor';
import { difficulty } from './stores/difficulty';
import { gamePaused } from './stores/game';
import { grid, userGrid } from './stores/grid';
import { timer } from './stores/timer';
import { hints } from './stores/hints';

/**
 * Start new game with a generated sudoku
 *
 * @param {('veryeasy' | 'easy' | 'medium' | 'hard')} diff - Difficulty
 */
export function startNew(diff) {
	difficulty.set(diff);
	grid.generate(diff);
	cursor.reset();
	timer.reset();
	hints.reset();

	location.hash = '';
}

/**
 * Start new game with a custom sudoku
 *
 * @param {string} sencode - Sencode to decode
 */
export function startCustom(sencode) {
	difficulty.setCustom();
	grid.decodeSencode(sencode);
	cursor.reset();
	timer.reset();
	hints.reset();
}

/**
 * Pause the game
 */
export function pauseGame() {
	timer.stop();
	gamePaused.set(true);
}

/**
 * Resume (un-pause) the game
 */
export function resumeGame() {
	timer.start();
	gamePaused.set(false);
}

export function undoGame() {
	userGrid.undo();
}

export function redoGame() {
	userGrid.redo();
}

export function canUndoGame() {
	return userGrid.canUndo();
}

export function canRedoGame() {
	return userGrid.canRedo();
}

export function startExplore() {
	userGrid.startExplore();
}

export function commitExplore() {
	userGrid.commitExplore();
}

export function cancelExplore() {
	userGrid.cancelExplore();
}

export function backtrackExplore() {
	userGrid.backtrackExplore();
}

export default {
	startNew,
	startCustom,
	pause: pauseGame,
	resume: resumeGame,
	undo: undoGame,
	redo: redoGame,
	startExplore,
	commitExplore,
	cancelExplore,
	backtrackExplore
};
